File: cli/test/receipts_check.py
#!/usr/bin/env python3
# 回执（已送达/已读）端到端回归：两台 TUI（A 创建、B 加入）互发一条，
# B 收到后应把"已读"高水位上报到服务端，且落盘到自己的 store。
#
# 语义（与服务端一致）：B 上报 read_upto_seq=N ⟺ B 已读到 seq ≤ N 的对方消息；
# 读隐含送达（服务端 SQL 保证 delivered ≥ read）。
#
# 运行：python3 cli/test/receipts_check.py
import json, os, pty, subprocess, select, time, socket, tempfile, urllib.request, sys

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
CLI = os.path.join(ROOT, 'cli')
WORK = tempfile.mkdtemp(prefix='einz-receipts-')


def free_port():
    s = socket.socket()
    s.bind(('127.0.0.1', 0))
    port = s.getsockname()[1]
    s.close()
    return port


def drain(master, seconds=1.0):
    out = b''
    deadline = time.time() + seconds
    while time.time() < deadline:
        r, _, _ = select.select([master], [], [], 0.2)
        if r:
            try:
                d = os.read(master, 8192)
            except OSError:
                break
            if not d:
                break
            out += d
    return out.decode('utf-8', errors='replace')


def send(master, s):
    os.write(master, s.encode())


def wait_text(master, text, timeout=40):
    out = ''
    deadline = time.time() + timeout
    while text not in out and time.time() < deadline:
        out += drain(master, 0.8)
    return out


def start_tui(store, port):
    master, slave = pty.openpty()
    env = dict(os.environ, TERM='xterm-256color')
    p = subprocess.Popen(
        ['dart', 'run', 'bin/einz_tui.dart', '--store', store,
         '--server', f'http://127.0.0.1:{port}'],
        stdin=slave, stdout=slave, stderr=slave, close_fds=True, cwd=CLI, env=env)
    os.close(slave)
    return master, p


def read_store(path):
    with open(path) as f:
        return json.load(f)


def wait_store_field(path, field, minimum=1, timeout=30, debug=False):
    """轮询 store JSON：等某字段 ≥ minimum（回执上报是异步的）。"""
    deadline = time.time() + timeout
    while time.time() < deadline:
        try:
            st = read_store(path)
            val = st.get(field, 0) or 0
            if debug:
                print('   [poll] %s=%s keys=%s' % (
                    field, st.get(field, '<missing>'),
                    'has' if field in st else 'MISSING'), flush=True)
            if val >= minimum:
                return val
        except Exception as e:
            if debug:
                print('   [poll] read error: %r' % (e,), flush=True)
        time.sleep(5)
    return 0


def wait_ws_connected(port, store_path, timeout=60):
    """等本设备的 WS 连上（服务端 devices.connected_at 非空 = 有实时 WS 连接）。"""
    st = read_store(store_path)
    deadline = time.time() + timeout
    while time.time() < deadline:
        try:
            devs = http_json(f'http://127.0.0.1:{port}/devices',
                             token=st['session_token'])['devices']
            for d in devs:
                if d.get('device_id') == st.get('device_id') and d.get('connected_at'):
                    return True
        except Exception:
            pass
        time.sleep(2)
    return False


def http_json(url, method='GET', body=None, token=None):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('Content-Type', 'application/json')
    if token:
        req.add_header('Authorization', f'Bearer {token}')
    with urllib.request.urlopen(req, timeout=10) as r:
        return json.load(r)


def main():
    port = free_port()
    server = None
    spawned = []
    try:
        server = subprocess.Popen(
            ['node', os.path.join(ROOT, 'server', 'dist', 'app.js')],
            env={**os.environ, 'PORT': str(port),
                 'EINZ_DB': os.path.join(WORK, 'einz.sqlite.db'),
                 'EINZ_FILES': os.path.join(WORK, 'files')},
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        for _ in range(50):
            try:
                urllib.request.urlopen(f'http://127.0.0.1:{port}/health', timeout=1)
                break
            except Exception:
                time.sleep(0.2)

        # ---------- A：创建秘境 ----------
        store_a = os.path.join(WORK, 'a.json')
        ma, pa = start_tui(store_a, port)
        spawned.append((ma, pa))
        for expect, payload in [
            ('秘境入口', 'c\r'), ('我的名字', 'Lukas\r'), ('我的性别', '1\r'),
            ('伴侣的名字', 'Alice\r'), ('伴侣的性别', '2\r'),
        ]:
            out = wait_text(ma, expect)
            if expect not in out:
                print(f'❌ A 未到「{expect}」问答'); print(out[-600:]); return 1
            send(ma, payload)
        out = wait_text(ma, '设置密保口令')
        if '设置密保口令' not in out:
            print('❌ A 未到口令问答'); print(out[-600:]); return 1
        send(ma, 'abc123\r')
        out = wait_text(ma, '成功创建秘境', timeout=40)
        if '成功创建秘境' not in out:
            print('❌ A 未创建成功'); print(out[-600:]); return 1
        # 锁屏码：TUI 阻塞读键盘时不重绘，提示要等下一次按键才出现 —— 用重试发送
        # （既有脚本同款 input_until；非法值会被拒，重发无害）
        def input_until(master, payload, expect, tries=8, timeout=8):
            out = ''
            for _ in range(tries):
                send(master, payload)
                out = wait_text(master, expect, timeout=timeout)
                if expect in out:
                    break
            return out

        out = input_until(ma, '123456\r', '🔢')
        if '🔢' not in out:
            print('❌ A 未设置锁屏码'); print(out[-800:]); return 1
        time.sleep(1)  # 等 store 落盘（space_id 需要）
        st_a = read_store(store_a)
        space_id = st_a.get('space_id')
        if not space_id:
            print('❌ A 的 store 缺少 space_id'); return 1

        # ---------- B：加入 ----------
        tok = http_json(f'http://127.0.0.1:{port}/spaces/{space_id}/join-tokens',
                        method='POST')['joinToken']
        store_b = os.path.join(WORK, 'b.json')
        mb, pb = start_tui(store_b, port)
        spawned.append((mb, pb))
        for expect, payload in [
            ('秘境入口', 'j\r'), ('输入邀请码', tok + '\r'), ('我是谁', 'Alice\r'),
        ]:
            out = wait_text(mb, expect)
            if expect not in out:
                print(f'❌ B 未到「{expect}」问答'); print(out[-600:]); return 1
            send(mb, payload)
        out = wait_text(mb, '验证密保口令')
        if '验证密保口令' not in out:
            print('❌ B 未到口令问答'); print(out[-600:]); return 1
        send(mb, 'abc123\r')
        out = wait_text(mb, '成功加入秘境', timeout=40)
        if '成功加入秘境' not in out:
            print('❌ B 未加入成功'); print(out[-600:]); return 1

        # B 也设锁屏码（同样需要重试发送）
        out = input_until(mb, '123456\r', '🔢')
        if '🔢' not in out:
            print('❌ B 未设置锁屏码'); print(out[-800:]); return 1
        time.sleep(2)
        # 等 B 的 WS 真的连上（服务端 devices.connected_at 非空 = 有实时连接）。
        # 不能用 TUI 标题栏的绿●判定——它也表示"对方在线"，会误判。
        if not wait_ws_connected(port, store_b, timeout=60):
            print('❌ B 的 WS 未连上（devices.connected_at 为空）'); return 1

        # ---------- A 发一条 → B 实时收到并上报已读 ----------
        # A 发一条：TUI 阻塞读键盘期间可能吞掉输入 —— 直到出现「已发送」确认
        # （重发无害：最多多发几条，seq 仍单调）
        out = ''
        for _ in range(5):
            send(ma, 'receipt-probe-1\r')
            out = wait_text(ma, '已发送', timeout=8)
            if '已发送' in out:
                break
        if '已发送' not in out:
            print('❌ A 未发出消息'); print(out[-800:]); return 1

        # 判定以**服务端**为准（store 落盘可能再晚一拍）：轮询 B 的回执行直到
        # read_upto_seq ≥ 1。B 的上报是异步的，且 TUI 的事件循环可能被阻塞读
        # 键盘拖住，窗口放宽到 180s。
        st_b = read_store(store_b)
        r = None
        rows = []
        deadline = time.time() + 180
        while time.time() < deadline:
            try:
                rows = http_json(f'http://127.0.0.1:{port}/receipts',
                                 token=st_b['session_token'])['receipts']
                mine = [x for x in rows if x['person_id'] == st_b['person_id']]
                if mine and mine[0]['read_upto_seq'] >= 1:
                    r = mine[0]
                    break
            except Exception:
                pass
            time.sleep(3)
        if r is None:
            print('❌ 服务端 B 的 read_upto_seq 未推进')
            print('  rows=%s' % (rows,))
            print('  B store: lastServerSequence=%s lastReportedDeliveredSeq=%s lastReportedReadSeq=%s'
                  % (read_store(store_b).get('last_server_sequence'),
                     read_store(store_b).get('last_reported_delivered_seq'),
                     read_store(store_b).get('last_reported_read_seq')))
            print('  --- B tail ---'); print(drain(mb, 2.0)[-3000:])
            return 1
        if r['delivered_upto_seq'] < r['read_upto_seq']:
            print(f'❌ 读隐含送达被破坏（delivered < read）：{r}'); return 1
        # store 落盘（防抖标记）最终也应对齐（成功才推进 → 最多晚一拍）
        read_seq = wait_store_field(store_b, 'last_reported_read_seq',
                                    r['read_upto_seq'], timeout=20)

        print(f'✅ 回执链路打通：B 已读高水位={r["read_upto_seq"]}、送达={r["delivered_upto_seq"]}（store 落盘 {read_seq}）')
        return 0
    finally:
        for m, p in spawned:
            if p.poll() is None:
                p.kill()
            try:
                os.close(m)
            except OSError:
                pass
        if server and server.poll() is None:
            server.kill()


if __name__ == '__main__':
    sys.exit(main())

File: cli/test/receipts_check.dart
// 回执（已送达/已读）端到端回归 —— 直接驱动 ChatSession，不走 TUI。
//
// 为什么不用 pty：TUI 的事件循环由键盘输入驱动，WS 投递/上报时机不确定，
// pty 版断言会偶发失败（见 worklog 2026-09-12）。这里改为直接构造两个
// ChatSession（共享同一个 spaceKey），对真实服务端跑一遍，完全确定性。
//
// 覆盖（老板 2026-09-12 定的语义）：
//  1) 收到对方消息后 sync → **上报已送达**（delivered_upto_seq = 本地锚点）
//  2) 补拉/同步**不**上报已读（read 只由 WS 实时到达推进）→ 此时服务端 read 应为 0
//  3) 服务端回读：delivered ≥ 1 且 delivered ≥ read（读隐含送达）
//
// 运行：dart run test/receipts_check.dart（需先 cd server && npm run build）
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:einz_shared/einz_shared.dart';
import 'package:einz_cli/chat_core.dart';
import 'package:einz_cli/store.dart';

const kPassphrase = 'pass123';

Future<int> _freePort() async {
  final s = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
  final port = s.port;
  await s.close();
  return port;
}

Future<Process> _startServer(int port, String workDir) async {
  final root = Directory.current.parent.path; // cli/ → 仓库根
  final p = await Process.start(
    'node',
    ['${root}/server/dist/app.js'],
    environment: {
      ...Platform.environment,
      'PORT': '$port',
      'EINZ_DB': '$workDir/einz.sqlite.db',
      'EINZ_FILES': '$workDir/files',
    },
  );
  // 等就绪
  final client = HttpClient();
  for (var i = 0; i < 60; i++) {
    try {
      final req = await client.getUrl(Uri.parse('http://127.0.0.1:$port/health'));
      final res = await req.close();
      await res.drain<void>();
      client.close();
      return p;
    } catch (_) {
      await Future<void>.delayed(const Duration(milliseconds: 200));
    }
  }
  client.close();
  throw StateError('服务端未就绪');
}

DeviceStore _store({
  required String publicKey,
  required String privateKey,
  required String spaceKeyB64,
  required String spaceId,
  required String deviceId,
  required String personId,
  required String sessionToken,
  required String server,
}) {
  final st = DeviceStore(
    publicKey: publicKey,
    privateKey: privateKey,
    spaceKey: spaceKeyB64,
    spaceId: spaceId,
    keyVersion: 1,
    sessionToken: sessionToken,
  );
  st.deviceId = deviceId;
  st.personId = personId;
  return st;
}

Future<Map<String, dynamic>> _getReceipts(String server, String token) async {
  final client = HttpClient();
  try {
    final req = await client.getUrl(Uri.parse('$server/receipts'));
    req.headers.set(HttpHeaders.authorizationHeader, 'Bearer $token');
    req.headers.set('X-Protocol-Version', '1'); // 协议硬校验（PROTOCOL.md §1）
    final res = await req.close();
    final text = await res.transform(utf8.decoder).join();
    if (res.statusCode != 200) {
      throw StateError('GET /receipts -> ${res.statusCode} $text');
    }
    return jsonDecode(text) as Map<String, dynamic>;
  } finally {
    client.close(force: true);
  }
}

Future<int> _run() async {
  final work = await Directory.systemTemp.createTemp('einz-rcpt-');
  final port = await _freePort();
  final server = 'http://127.0.0.1:$port';
  final proc = await _startServer(port, work.path);

  try {
    final api = ApiClient(server);
    // A 创建空间（sealedSpaceKey 仅需结构合法；本测试不跑 escrow 解密流程）
    final created = await api.createSpace(
      publicKey: 'pk-a',
      personName: 'Lukas',
      partnerName: 'Alice',
      escrowPassphrase: kPassphrase,
      sealedSpaceKey: PassphraseEnvelope(
        salt: Uint8List.fromList(List.filled(16, 1)),
        nonce: Uint8List.fromList(List.filled(24, 2)),
        ciphertext: Uint8List.fromList(List.filled(8, 3)),
      ),
    );
    // B 凭 join token 加入（拿到 space 级 session）
    final joined = await api.joinSpace(
      token: created.joinToken,
      publicKey: 'pk-b',
      partnerSlot: 1,
      deviceName: 'b',
    );

    // 两端共享同一个 spaceKey（真实流程里由 escrow 口令包传递，这里直接给定——
    // 回执链路与密钥分发无关）
    final spaceKey = Uint8List.fromList(
        List<int>.generate(32, (_) => Random.secure().nextInt(256)));
    final keyB64 = base64Encode(spaceKey);

    final storeA = _store(
      publicKey: 'pk-a',
      privateKey: 'sk-a',
      spaceKeyB64: keyB64,
      spaceId: created.spaceId,
      deviceId: created.deviceId,
      personId: created.creatorPersonId,
      sessionToken: created.sessionToken,
      server: server,
    );
    final storeB = _store(
      publicKey: 'pk-b',
      privateKey: 'sk-b',
      spaceKeyB64: keyB64,
      spaceId: joined.spaceId,
      deviceId: joined.deviceId,
      personId: joined.personId,
      sessionToken: joined.sessionToken,
      server: server,
    );

    final sessionA = ChatSession(storeA, '${work.path}/a.json', server);
    final sessionB = ChatSession(storeB, '${work.path}/b.json', server);

    // A 发一条
    final ok = await sessionA.sendText('receipt-probe');
    if (!ok) {
      stderr.writeln('❌ A 发送失败');
      return 1;
    }

    // B 同步（补拉）→ 应上报"已送达"、**不**上报已读
    await sessionB.sync();
    if (storeB.lastReportedDeliveredSeq < 1) {
      stderr.writeln('❌ B 未上报已送达（lastReportedDeliveredSeq='
          '${storeB.lastReportedDeliveredSeq}）');
      return 1;
    }
    if (storeB.lastReportedReadSeq != 0) {
      stderr.writeln('❌ B 的补拉不应上报已读（lastReportedReadSeq='
          '${storeB.lastReportedReadSeq}）');
      return 1;
    }

    // 服务端回读
    final receipts = (await _getReceipts(server, storeB.sessionToken!))['receipts']
        as List<dynamic>;
    final row = receipts.cast<Map<String, dynamic>>().firstWhere(
          (r) => r['person_id'] == storeB.personId,
          orElse: () => <String, dynamic>{},
        );
    if (row.isEmpty) {
      stderr.writeln('❌ 服务端查不到 B 的回执行：$receipts');
      return 1;
    }
    final delivered = row['delivered_upto_seq'] as int;
    final read = row['read_upto_seq'] as int;
    if (delivered < 1) {
      stderr.writeln('❌ 服务端 delivered_upto_seq 未推进：$row');
      return 1;
    }
    if (delivered < read) {
      stderr.writeln('❌ 读隐含送达被破坏（delivered < read）：$row');
      return 1;
    }
    if (read != 0) {
      stderr.writeln('❌ 补拉后服务端 read 应为 0（read 只由 WS 实时到达推进）：$row');
      return 1;
    }

    stdout.writeln('✅ CLI 回执链路：B 补拉上报已送达=$delivered、已读=$read'
        '（补拉不标已读 ✓）');
    return 0;
  } finally {
    proc.kill();
    try {
      await work.delete(recursive: true);
    } catch (_) {}
  }
}

Future<void> main() async {
  await sodium();
  final code = await _run();
  exit(code);
}
